components/feature/recentlyclosed/src/main/java/mozilla/components/feature/recentlyclosed/RecentlyClosedMiddleware.kt

```diff
@@ -0,0 +1,118 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.feature.recentlyclosed
+
+import android.content.Context
+import androidx.annotation.VisibleForTesting
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.flow.collect
+import kotlinx.coroutines.launch
+import mozilla.components.browser.state.action.BrowserAction
+import mozilla.components.browser.state.action.RecentlyClosedAction
+import mozilla.components.browser.state.action.TabListAction
+import mozilla.components.browser.state.selector.findTab
+import mozilla.components.browser.state.selector.normalTabs
+import mozilla.components.browser.state.state.BrowserState
+import mozilla.components.browser.state.state.ClosedTab
+import mozilla.components.lib.state.Middleware
+import mozilla.components.lib.state.MiddlewareContext
+import mozilla.components.lib.state.Store
+
+/**
+ * [Middleware] implementation for handling [RecentlyClosedAction]s and syncing the closed tabs in
+ * [BrowserState.closedTabs] with the [RecentlyClosedTabsStorage].
+ */
+class RecentlyClosedMiddleware(
```

Comment:
Yeah for now this is not ideal behavior 😕 but I think it's acceptable until we move the undo action into AC. Hacking something into the Fenix undo actions doesn't seem like a great option right now either.

Reply:
Yeah, I agree.

I started to prototype a patch that will add undo support to AC. It will have the some of the same problems since we still need to support SessionManager. But then we can easily move it over to BrowserStore and improve it once the time comes.

Reply:
But the advantage is going to be that we can only fill the list of recently closed tabs once the undo has timed out. I'll ping you on the PR once I have something. :)